---
title: "Class 11"
author: "Gary Le"
date: "May 9, 2018"
output: github_document
---

# Structural Bioinformatics

## Notes

* Paradigm: Sequence > Structure > Function
    + Structure determines Function
    + Structure is more conserved than sequence (e.g. viral proteins)
    + But, structure is encoded by sequence

* Energy Landscape
    + shows the energy as a molecule over time.
    + used to show the valleys where stable conformations exist
  
* Motivations
    + Detailed understanding of molecular interactions; understand how the system is perturbed
    + Theoretical and computational predictions to open new ideas for confirmation (e.g. DNA)
  
* Goals:
    + Analyze
    + Visualize
    + Compare
    + Predict
    + Design
    
* Review:
    + Secondary structures are brought out by interactions between residues and by Phi and Psi bond rotations
    + Alpha Helix:
        + Stabilized by "i and "i + 4"" residue interactions 
    + Proteins are surrounded by water and other solvents, not empty space
    
## Hand-ons Practice

### Section 1:

Q1: Download a CSV file from the PDB site (accessible from “Analyze” -> “PDB Statistics” >
“by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project
and determine the percentage of structures solved by X-Ray and Electron Microscopy. From
the website what proportion of structures are protein?

```{r}
pdb.csv <- read.csv("Data Export Summary.csv", row.names = 1)

#Finding percentage of proteins resolved by each method
pdb.percent <- pdb.csv$Total/ sum(pdb.csv$Total) * 100

#Conserving row names
names(pdb.percent) <-  row.names(pdb.csv)

#Output answer
print(pdb.percent["X-Ray"])

print(pdb.percent["Electron Microscopy"])

```
Overwhelmingly, most protein structures are resolved by X-Ray crystallography as expected.

```{r}
#Totaling numbers of proteins resolved by all methods
sum(pdb.csv$Proteins)
```


Q2: Type HIV in the PDB website search box on the home page and determine how many
HIV-1 protease structures are in the current PDB?
793 structures

### Section 3: Bio3D Visualization

```{r}
#Loading library
library("bio3d")
```

```{r}
pdb <- read.pdb("1hsg.pdb")

#Protein summary of HIV-1 Protease
print(pdb)
```
Q6. How many amino acid residues are there in this pdb object and what are the two nonprotein
residues?

* 198 residues in the protein
* The two non-protein residue types are water and the ligand MK1

```{r}
#Retrieving b-factor of alpha carbons
pdb.calpha.bfac <-  pdb$atom$b[pdb$calpha]

#Plotting beta-factors
plot.bio3d(pdb.calpha.bfac, sse=pdb, typ="l", ylab = "B-Factor")

```

Trying to select only the protein atoms for viewing with VMD

* Finding indices from the pdb file
```{r}
#Using Bio3D to find the indices of only the protein
inds.p <- atom.select(pdb, "protein")

#Finding indices for ligand
inds.l <- atom.select(pdb, "ligand")
```

* Using indices to subset the main pdb file
```{r}
#Showing protein atoms
pdb$atom[inds.p$atom,]
```

```{r}
#Showing ligand atoms
pdb$atom[inds.l$atom,]
```

Using the trim.pdb function to make a truncated PDB object
```{r}
#Creating PDB object w subset atoms for ligand
pdb.ligand <- trim(pdb, inds = inds.l)
pdb.ligand

#Creating PDB object w subset atoms for protein
pdb.protein <- trim(pdb, inds = inds.p)

#Writing Files
write.pdb(pdb.ligand, file = "1hsg_ligand") #Ligand file
write.pdb(pdb.protein, file = "1hsg_protein") # Protein file
```

### Section 4:

* MUSCLE downloaded and can be used by R to align PDB files locally

```{r}
# Download some example PDB files
ids <- c("1TND_B","1AGR_A","1TAG_A","1GG2_A","1KJY_A","4G5Q_A")
files <- get.pdb(ids, split = TRUE) #Getting files and splitting by chains

#PDB files are downloaded and subdirectory created to store chains.
```

* Using pdbaln() to align PDB files
```{r}
# 'fit = TRUE' tells rmsd() to complete super imposition calculations
pdbs.aln <- pdbaln(files = files, fit = TRUE, exefile = "C:/Users/gary_/Desktop/R_Bootcamp/bggn213_github/class11/muscle.exe")
```

### Section 5: Example Application on Adenylate Kinase (Adk)

* Retrieving Adk seq from pdb ID
```{r}
#Amino acid sequence for chain A
aa <- get.seq("1ake_A")
```

* BLAST search the sequence and store results
```{r}
#Blast results
blast <- blast.pdb(aa)

```

* Plotting BLAST results

```{r}
hits <- plot(blast)
```

```{r}
#Stores PDB id of blast hits among other things
hits
```

* Using blast hits to retrieve corresponding pdb files
```{r}
#Storing PDB files in dir "pdbs"
files.adk <- get.pdb(hits$pdb.id, path = "pdbs", split = TRUE, gzip =
TRUE)

#Aligning BLAST hits to ADK
pdbs.adk <- pdbaln(files = files.adk, fit = TRUE, exefile = "C:/Users/gary_/Desktop/R_Bootcamp/bggn213_github/class11/muscle.exe")
```

* Graphing alignment
```{r}
# Vector containing PDB codes; Used to label alignment
ids <- basename.pdb(pdbs.adk$id)

# Draw schematic alignment
plot(pdbs.adk, labels=ids)
```
We clearly see three groups of protein structures. They should show the different conformations of the proteins.

* Calculating sequence conservation by Shannon Entropy by Residue
```{r}
# Calculate sequence conservation
cons <- conserv(pdbs.adk, method="entropy22")

# SSE annotations
sse <- pdbs2sse(pdbs.adk, ind=1, rm.gaps=FALSE)

# Plot conservation per residue
plotb3(cons, sse=sse, ylab="Sequence entropy")
```

* Installing required packages for next step.
```{r}
install.packages("XML")
install.packages("RCurl")

library("XML")
library("RCurl")
```


```{r}
#Annotating the PDB files of the blast hits
## N.B. Requires the XML and RCurl packages
anno <- pdb.annotate(ids)

#Printing out the unique source species
print(unique(anno$source))
```

```{r}
# Perform PCA & plot the results
pc.xray <- pca(pdbs.adk)
plot(pc.xray)
```

* Performing PCA on ADK groups and coloring H-cluster appropriately.
```{r}
# Calculate RMSD
rd <- rmsd(pdbs.adk)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)
```

